### PR TITLE
Features/wasm components

### PR DESCRIPTION
**Pull request by superchris**
This is a draft PR to show how the progress is coming along for supporting web assembly components. Check out `wasm_components_test.exs` to see where things are at

### CODE REVIEW

**Comment by superchris** [comment]
So I've hit a bit of wall here where it looks I'll need to do something drastic to get component support working the way I think it should. A little conext: A wasm component has, as it's definition, a `.wit` file which specifies the high level, typed, interface to the component. The `wasmtime` rust crate provides a `bindgen` macro which generates the rust bindings for a component. This means that for a component to be used, their will need to be some project specific rust code which depends on the the wasmex rust code. Trying to have one rustler nif depend on another rustler nif did not seem to go very well, which makes me think what would need to happen is the rustler nif would need to be specific to the components being used. In order for this to be viable, there would need to be considerable change to the wasmex codebase methinks, probably involving allowing the use to specify the `Native` module via configuration so that they could sub in their own, and then maybe a `use Wasmex.Native` macro to bring in all the wasmex stuff. I don't know if I've described this well enough to make any sense at all, but I'm trying to understand if I should keep marching in this direction or if this is so radical that I should fork and do component specific hex package. Love to hear any thoughts you might have @tessi etc

**Comment by tessi** [comment]
Hey @superchris

first things first: it's awesome that you start tackling this. Having good component support is the next most pressing thing for wasmex to have. thank you 💜

I started looking into components but didn't had the time yet to start a real prototype - so I'm afraid my understanding of how components work isn't perfect yet. My (maybe naïve?) understanding of components was that they are more or less a normal wasm modules with added .wit type definitions. Those type definitions could be "used" by wasmex to read/write function results/params. I wasn't sure how that "using" of the WIT types could be done - my hope was to maybe have an elixir macro that defines proper elixir-structs, which wasmex then knows how to (de-)serialise to what the bytes the wasm module needs. Basically, like building our own bindgen.

Coming from that understanding, I didn't get the part on why we need to link other NIFs or use the wasmtime bindgen. Can you elaborate there, please? I probably just don't know enough to fully understand the problem - so just dropping a link to the relevant docs for me to educate myself also works.

If you're up for it, we could also just have a call and discuss/brainstorm possible approaches synchronous. :)

**Comment by superchris** [comment]
Hey @tessi! Thanks for wasmex :) So the issue I am facing is that I will need to generate two kinds of bindings, one in rust, and another in Elixir. In my branch I the rust bindings I need are in `native/wasmex/src/todo_list.rs` and the elixir bindings are the `todo_*` functions in `Native`. These will need to exist for each component, at least the way I understand things right now. This means each project will have rust code of it's own that will depend on the core wasmex rust code, which is the wall I hit. I'd be delighted to do a brainstorm call (and maybe try to better explain things), feel free to suggest some times that are good for you. I'm UTC-4 (EDT).

**Comment by superchris** [comment]
@tessi I think after reading your comments again I understand what you are saying. Having our own Elixir bindgen would be amazing, I was trying to do a lazier approach because I was intimidated by making to do the lowering and lifting of elixir types into wasm memory. I probably am not facile enough with rust and the internals of wasm to be able to tackle the elixir bindgen idea just yet. Anyways, happy to talk if you'd like 😄

**Comment by tessi** [comment]
brainstorm idea: if we can implement `Lower` and `Lift` for elixir types/structs (see: https://github.com/bytecodealliance/wasmtime/blob/e51a68e64ae2e955b69a6f9e37f5479f8186ae91/crates/wasmtime/src/runtime/component/func/typed.rs#L505), then we have the load+store operations done. 🤔 

wasmtimes bindgen! macro basically utilizes the Lift/Lower traits, which we can implement ourselves for elixir-structs instead of deriving it automatically from known rust types.

**Comment by superchris** [comment]
Hey @tessi I got a little further in my implementation and at this point I've hacked things up to the point its a bit of a mess ;) I had to switch out the wasi library to new ones (`wasmtime-wasi` and `wasmtime-wasi-http`) that support WASI p2. As a result, I was able to call into a component in JS that did external http calls using JS native fetch, which was pretty exciting. However, I'm still not quite sure how to move forward with an approach that wouldn't involve per component rust code. I think I might spin up a component specific package, even if just temporarily, to see what using per component rust code would even look like. However, I would be happy to pair on your ideas for wasmex and pursue both approaches.

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
+  "rust-analyzer.linkedProjects": [
+    "./native/wasmex/Cargo.toml"
+  ]
+}
```
Comment (tessi) [inline .vscode/settings.json:5]: uhh, this is a good one! will shamelessly copy that over to `main`

**Comment by superchris** [comment]
I did end up making some amount of progress after I decided to split off to another project at least temporarily. It's still got some hand coded plumbing I need to get rid of, but have several examples working with tests to at least prove out the approach. Code is here: https://github.com/launchscout/wasm_components_ex

However after looking at [wasmtime-rb](https://github.com/bytecodealliance/wasmtime-rb) where the beginning of component support has landed, I think their approach is better and would allow us to dynamically invoke functions on components without a separate nif lib per component as my current approach requires. I may try another PR based on it. In short, they convert ruby types back and forth to wasmcomponent val types in rust. This seems like a much smaller lift (pun intended :) ).

I still think the lift/lower idea is a good one, just not quite sure how to start and how ambitious it is compared to the other approaches. Happy to pair some time @tessi if you'd like

**Comment by tessi** [comment]
wow, cool! will have a look this evening - both, your examples as well as wasmtime-rb

actually, i'm not too too set on a single idea as long as users don't have to learn rust/nifs (given they already have a .wasm component). I mean, if that's what's required we have to, but better not :)

re pairing: I'm sure by now you're the subject matter expert on components :) happy to pair though. Let's arrange something in a private channel. How can I reach you best? bluesky/mastodon would be my favourites, mail works too

**Comment by superchris** [comment]
Hey @tessi I just started trying to use BlueSky more I am @superchrisnelson.bsky.social

**Comment by superchris** [comment]
That sounds awesome more experiments are the way to go. Eventually the best path forward will emerge :)

**Comment by superchris** [comment]
@tessi I actually made some progress and was able to dynamically invoke a component function. I think it may be just a matter of filling in all the type conversions from here (not to minimize that effort, which isn't trivial).

File: test/support/hello_world/hello-world.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  var results = [];
+  for(let i = 0; i < times; i++) {
+    results.push(greet(who));
+  }
+  return results;
+}
+
+export function greetMany(people) {
+  return people.map((person) => `Hello, ${person}!`);
+}
```
Comment (tessi) [inline test/support/hello_world/hello-world.js:15]: oh we have js now! what's the purpose of this file?

Comment (tessi) [reply test/support/hello_world/hello-world.js:15]: ah wait. this is the source code of the hello-world wasm component?!

File: test/wasm_components_test.exs [hunk 3]
```diff
@@ -0,0 +1,43 @@
+defmodule Wasmex.WasmComponentsTest do
+  use ExUnit.Case, async: true
+
+  alias Wasmex.Engine
+  alias Wasmex.EngineConfig
+
+  test "invoke component func" do
```
Comment (tessi) [inline test/wasm_components_test.exs:7]: regarding test setup, I'd love to see one (bigger) wasm component that tests all (or at least most) input/output variations with all types. Something similar to what wasmtime-ruby does or what my PR does in the test setup.

However, I love your descriptive examples. Maybe we can choose *one* "good" example to show-off component support in a test. what do you think? Feel free to steal the test setup in my MR if you want.

File: lib/wasmex/component_store.ex [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    %__MODULE__{
+      resource: resource,
+      reference: make_ref()
+    }
+  end
+
+  def new(%WasiOptions{} = options, store_limits \\ nil, engine \\ nil) do
+    %Engine{resource: engine_resource} = engine || Engine.default()
+
+    case Wasmex.Native.store_new_wasi_p2(
```
Comment (tessi) [inline lib/wasmex/component_store.ex:17]: hmm, do we want to separate wasi and non-wasi components maybe? or do all wasm components need to have wasi enabled?

File: native/wasmex/src/exec_func.rs [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+    // let paramTypes = func.params(store)
+    // .ty(&*component_store)
+    // .params()
+    // .collect::<Vec<ValType>>();
+    let paramTypes = func.params(&mut *component_store);
+    let converted_params = convert_params(&mut *component_store, paramTypes, given_params)?;
+    println!("Converted params: {:?}", converted_params);
+
+    let list = vec![wasmtime::component::Val::String(String::from(""))];
```
Comment (tessi) [inline native/wasmex/src/exec_func.rs:64]: whydo we need this list?

File: test/support/hello_world/hello-world.js (hunk 2 above)
Comment (superchris) [reply test/support/hello_world/hello-world.js:15]: yes :)

File: lib/wasmex/component_store.ex (hunk 4 above)
Comment (superchris) [reply lib/wasmex/component_store.ex:17]: I refactored this to have a WASIP2Options. I *think* all components need some amount of WASI enabled, but the options are different. Lemme know if you think this approach is good

File: native/wasmex/src/exec_func.rs (hunk 5 above)
Comment (superchris) [reply native/wasmex/src/exec_func.rs]: we don't is gone :)

File: test/wasm_components_test.exs (hunk 3 above)
Comment (superchris) [reply test/wasm_components_test.exs:7]: I added a component types test by stealing more from wasmtime-rb :) Their example was built in rust not javascript so it is waaay faster to boot. I may or may not keep the original around. It probably makes sense to move it out of a test into an example directory or something. Because it was implemented in js it will always be super slow to load

File: lib/wasmex/component_instance.ex [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+      resource -> {:ok, __wrap_resource__(store_or_caller_resource, resource)}
+    end
+  end
+
+  def call_function(
+        %__MODULE__{store_resource: store_resource, instance_resource: instance_resource},
+        function,
+        args
+      ) do
+    {:ok, Wasmex.Native.exec_func(store_resource, instance_resource, function, args)}
```
Comment (tessi) [inline lib/wasmex/component_instance.ex:34]: * I think it's good to have `exec_fun` separate from `instance_call_exported_function`, but both are very similar things. I would rename both to have the consistent naming (e.g. `instance_call_function` and `component_call_function`). What do you think? No need to rename `instance_call_exported_function` in this PR (can be a follow-up) but the new call-function name can starts in a good place when we just introduce it 🤔

File: lib/wasmex/native.ex [hunk 7]
```diff
@@ -5,23 +5,9 @@ defmodule Wasmex.Native do
...
-      riscv64gc-unknown-linux-gnu
-      x86_64-apple-darwin
-      x86_64-pc-windows-gnu
-      x86_64-pc-windows-msvc
-      x86_64-unknown-freebsd
-      x86_64-unknown-linux-gnu
-      x86_64-unknown-linux-musl
-    ),
-    force_build: System.get_env("WASMEX_BUILD") in ["1", "true"]
+    mode: :debug
```
Comment (tessi) [inline lib/wasmex/native.ex:10]: I imagine you do this for local development, right? in the finished PR I'd remove this change (if this is a proposal, let's maybe move it to a separate PR)

File: lib/wasmex/component_instance.ex (hunk 6 above)
Comment (superchris) [reply lib/wasmex/component_instance.ex:34]: Agreed, renaming this was one of the next things on my list. I'd like to keep it consistent as well, at first I was just thinking I'd rename mine to `component_instance_call_exported_function` but that is a mouthful :) I'm good with `component_call_function`

File: lib/wasmex/native.ex (hunk 7 above)
Comment (superchris) [reply lib/wasmex/native.ex:10]: Yep, just leftover just from development :) Will remove

File: test/wasm_components_test.exs (hunk 3 above)
Comment (tessi) [reply test/wasm_components_test.exs]: sounds good. I really like the approach wasmtime-rb did in their testing. very systematic and easy to see what's supported.

regarding an example: I'm happy to have one more wasm_component that is a tiny bit more complex (probably something using wasi, which the wasmtime-rb test does not use to my knowledge)

All other examples, IMO should live in separate repositories. Reason is that they should either:

1. live in this repo (in an example repo), but then we need to test them so they don't get stale
2. or live in example repository/repositories - which can test on their own and are way better "templates" is someone wants to start a project

To me 2. looks like having more benefits with less disadvantages for the main repo

File: lib/wasmex/component_store.ex (hunk 4 above)
Comment (tessi) [reply lib/wasmex/component_store.ex]: hmm, not sure. I need to research better but my understanding so far is that wasi-p2 is just a couple of different worlds a component can depend on and wasmtime gives a default implementation if we enable it to do so. But we could totally tell wasmtime to NOT enable any WASI P2 and instead implement the world in Elixir.

I want to have the flexibility to do that and have a default which is 100% sandboxes. There is some security and usability background to it. I want users of wasmex to explicitly and knowingly enable API surface to their components. Nobody should be surprised that an attacker could suddenly read from stdin, do HTTP calls, get the current time or anything else potentially security sensitive operation.

File: lib/wasmex/components.ex [hunk 8]
```diff
@@ -0,0 +1,39 @@
+defmodule Wasmex.Components do
```
Comment (tessi) [inline lib/wasmex/components.ex:1]: interesting, do you plan to have a different GenServer than our default Wasmex one? What's the reasoning?
Do you plan to eventually go full-component and remove the default one? Do you think that over time we need more specialised GenServers? Do you think they are implemented so different that they do not have much overlap?

I'm open to ideas, but want to understand the reasoning :) My first instinct is that this and the old GenServer will probably look very much the same and end up implementing almost the same thing. Maybe they should be the same? But then, if every handle_call is a big switch, we might end up very spaghetti and it's better to separate both things (plain wasm, components) from the start 🤔  i'm very interested in your thoughts

Comment (tessi) [reply lib/wasmex/components.ex:1]: uhhh, while we talk components. There was a very helpful comment by José a while back where he suggested our GenServer should queue incoming requests in case the client is busy running some WASM.

https://github.com/tessi/wasmex/issues/391#issuecomment-1374279165

maybe we should implement that idea from the start in this new GenServer (but maybe in a separate PR, if you prefer)

File: lib/wasmex/component_store.ex (hunk 4 above)
Comment (superchris) [reply lib/wasmex/component_store.ex]: That makes sense, I very much agree we don't want to give the user a foot gun. I'll do some experiments and see if things still work with WASI disabled. If not we can talk about what to do.

File: lib/wasmex/components.ex (hunk 8 above)
Comment (superchris) [reply lib/wasmex/components.ex:1]: I initially went this direction based on what I saw in the wasmtime codebase, where they have separate packages for components and non-component wasm. I also need a different version of store, and there is a different version WASI. It might be possible to merge into a single GenServer module, but the more I look at the code I'm less confident of trying do this. Keeping them separate also lets me get something out while I'm gradually adding features (linking, etc) that you already have in the main `Wasmex` genserver.

Long term, I like the idea of allowing the user to define component specific genservers and building macros to give them more specific, high-level functions from the wit definitions. Eg, a `TodoList` genserver module would have an `add_todo` fuction which took a pid and the arguments to pass to the exported function, with the correct arity rather than having to bundle the args into a list.

For the queuing idea I kinda of want to wait on that. It strikes me as something we or someone else could layer in if they need it maybe?

File: test/wasm_components_test.exs (hunk 3 above)
Comment (superchris) [reply test/wasm_components_test.exs]: Yeah, i think moving the more complex components into separate examples repos is the way to go. This will keep are tests faster as well.

File: lib/wasmex/components.ex (hunk 8 above)
Comment (tessi) [reply lib/wasmex/components.ex:1]: 👍 on going separate
👍 on queueing later
and let's talk about the macro idea separate (my belly feeling points towards code generator and not towards macros, but i didn't explore this yet. so that's a very weak opinion)

Comment (superchris) [reply lib/wasmex/components.ex:1]: I don't have a strong opinion here, code generators vs macros. Might be fun to check out some of the work Ash has been doing with their Igniter thing. It might make sense for this to be a separate package even

File: native/wasmex/todo-list.wit [hunk 9]
```diff
@@ -0,0 +1,6 @@
+package local:todo-list;
+
+world todo-list {
+  export init: func() -> list<string>;
+  export add-todo: func(todo: string, todos: list<string>) -> list<string>;
+}
```
Comment (tessi) [inline native/wasmex/todo-list.wit:6]: is this todo-list still in use? (same for the other todo-list.wit file and the .wasm file)

File: native/wasmex/Cargo.toml [hunk 10]
```diff
@@ -12,14 +12,15 @@ edition = "2021"
 [lib]
 name = "wasmex"
 path = "src/lib.rs"
-crate-type = ["dylib"]
+crate-type = ["cdylib"]
```
Comment (tessi) [inline native/wasmex/Cargo.toml:15]: what is this change doing?

File: native/wasmex/todo-list.wit (hunk 9 above)
Comment (superchris) [reply native/wasmex/todo-list.wit]: nope just stuff left around when I was in "experiment" mode

File: native/wasmex/Cargo.toml (hunk 10 above)
Comment (superchris) [reply native/wasmex/Cargo.toml]: same as above

File: lib/wasmex/components/component_store.ex [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    case Wasmex.Native.component_store_new(
+           store_limits,
+           engine_resource
+         ) do
+      {:error, err} -> {:error, err}
+      resource -> {:ok, __MODULE__.__wrap_resource__(resource)}
+    end
+  end
+
+  def new(%WasiP2Options{} = options, store_limits, engine) do
```
Comment (tessi) [inline lib/wasmex/components/component_store.ex:28]: hmm, what do you think about having `new` and `new_wasi` functions here to mimick the non-component Stores API?

File: test/component_fixtures/hello_world/hello-world.js [hunk 12]
```diff
@@ -0,0 +1,15 @@
+export function greet(who) {
```
Comment (tessi) [inline test/component_fixtures/hello_world/hello-world.js:1]: more a FYI because I think we shouldn't tackle this in this PR: I'd love to change these js modules into Rust modules.

not so much because I don't like JS, but to reduce the number of different stacks we have to support in this Repo :)  I'm also hoping to reduce test runtime, but thats a side goal :)

File: lib/wasmex/components/component_store.ex (hunk 11 above)
Comment (superchris) [reply lib/wasmex/components/component_store.ex:28]: yeah, that probably makes sense. Consistency is good :)

File: test/component_fixtures/hello_world/hello-world.js (hunk 12 above)
Comment (superchris) [reply test/component_fixtures/hello_world/hello-world.js:1]: Yeah I agree, I was just too lazy to do it :) This one would probably be easy to do in rust, the http wasi example might be a little more effort.

**Comment by tessi** [comment]
@superchris this is great and I love to merge this (and iterate in separate PRs) 💜 
we just need to get linters happy first :)

**Comment by tessi** [comment]
Thanks again @superchris . This was a super cool feature for a first PR.  Can't wait to see what comes next :) 💜
